src/remote/utils.go

```diff
@@ -181,14 +211,32 @@ func (c *Client) actionURL(digest *pb.Digest, prefix bool) string {
 }
 
 // locallyCacheResults stores the actionresult for an action in the local (usually dir) cache.
-func (c *Client) locallyCacheResults(target *core.BuildTarget, digest *pb.Digest, metadata *core.BuildMetadata, ar *pb.ActionResult) {
+func (c *Client) locallyCacheResults(target *core.BuildTarget, dg *pb.Digest, metadata *core.BuildMetadata, ar *pb.ActionResult) {
 	if c.state.Cache == nil {
 		return
 	}
 	data, _ := proto.Marshal(ar)
 	metadata.RemoteAction = data
 	metadata.Timestamp = time.Now()
-	if err := c.mdStore.storeMetadata(digest.Hash, metadata); err != nil {
+
+	if len(ar.OutputDirectories) > 0 {
+		tree := pb.Tree{}
+		for _, d := range ar.OutputDirectories {
+			t := pb.Tree{}
+			if _, err := c.client.ReadProto(context.Background(), digest.NewFromProtoUnvalidated(d.TreeDigest), &t); err == nil {
+				tree.Children = append(tree.Children, t.Root)
+				tree.Children = append(tree.Children, t.Children...)
+			}
+		}
+		for _, dir := range tree.Children {
+			c.directories.Store(c.digestMessage(dir).Hash, dir)
+		}
+		directoriesStored.Add(float64(len(tree.Children)))
+		data, _ := proto.Marshal(&tree)
+		metadata.RemoteOutputs = data
+	}
+
+	if err := c.mdStore.storeMetadata(dg.Hash, metadata); err != nil {
```

Comment:
Would be interesting to see how much bigger the cache gets if we _are_ storing all of this in the local MD file. I can image that for stuff like the AWS SDK might be a bit chunky.

Reply:
```
$ du -sh ~/.cache/please/build-metadata-store/
47M	/home/pebers/.cache/please/build-metadata-store/
```

Seems fine. Even for something like the v1 AWS SDK I don't think the directory metadata is really all _that_ big.